operator: Reimplement service synchronization without service cache

joamaki opened the pull request:
The [k8s.ServiceCache] facility and [k8s.Service] type are going away with the switch to the new load-balancing control-plane. Since we don't need most of it for the k8s service to kvstore synchronization reimplement it from scratch.

joamaki committed:
operator: Reimplement service synchronization without service cache

The [k8s.ServiceCache] facility and [k8s.Service] type are going away with the switch
to the new load-balancing control-plane. Since we don't need most of it for the k8s
service to kvstore synchronization reimplement it from scratch.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+	Endpoints               resource.Resource[*k8s.Endpoints]
+	StoreFactory            store.Factory
+	ClusterServiceConverter ClusterServiceConverter
+	Backend                 promise.Promise[kvstore.BackendOperations] `optional:"true"`
```
Little bit messy doing it this way. These are not provided by operator's hive, but are provided in clustermesh-apiserver. Should probably look into switching the operator to using `kvstore.Cell`

giorio94 replied:
Finally managed to tackle that refactoring: https://github.com/cilium/cilium/pull/39927

joamaki commented on pkg/k8s/resource_ctors.go:
```diff
@@ -319,6 +320,15 @@ func EndpointsResource(logger *slog.Logger, lc cell.Lifecycle, cfg Config, cs cl
 		endpointsOptsModifiers:      append(opts, endpointsOptsModifier),
 		endpointSlicesOptsModifiers: append(opts, endpointSliceOptsModifier),
 	}
+	indexers := cache.Indexers{
```
Adding this indexing will add additional costs to cilium-agent. Perhaps should take additional indexers as arguments.

Long-term plan for cilium-agent is to get rid of `Resource[Service]` and `Resource[Endpoints]` and have `pkg/loadbalancer/reflectors` reflect directly into tables without separate `cache.Store`. That'll take some time though and won't make it to v1.18.

joamaki commented on operator/watchers/script_test.go:
```diff
@@ -0,0 +1,160 @@
+		"testdata/*.txtar")
+}
+
+func kvstoreListCommand(client kvstore.BackendOperations) script.Cmd {
```
Hmm, time to add this command to `kvstore.Cell` along with the `kvstore/update` and `kvstore/delete` from `pkg/clustermesh/script_test.go`?

giorio94 replied:
Moved in https://github.com/cilium/cilium/pull/39965

joamaki commented on operator/watchers/script_test.go:
```diff
@@ -0,0 +1,160 @@
+
+		h := hive.New(
+			client.FakeClientCell,
+			daemonk8s.ResourcesCell,
```
FIXME:  don't use `daemon/k8s` here

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -284,7 +284,9 @@ func (c *inMemoryClient) LockPath(ctx context.Context, path string) (KVLocker, e
-	panic("unimplemented")
+	if fn != nil {
+		fn(prefix)
+	}
```
This implementation is not correct, because `fn` should be called only if a lease for that prefix expired. In this case, it is safe to just do nothing, given that there are no leases.

joamaki replied:
Ah yes, fixed!

giorio94 commented on operator/watchers/testdata/servicesync.txtar:
```diff
@@ -0,0 +1,151 @@
+hive/start
```
The `cluster-id` and `cluster-name` parameters should be set to non-default values.

joamaki replied:
Good catch, fixed!

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+func registerServiceSync(jg job.Group, p ServiceSyncParams) {
+	if !operatorOption.Config.SyncK8sServices || option.Config.KVStore == "" || !p.Clientset.IsEnabled() {
```
Nit. `operatorOption.Config.SyncK8sServices` is not set in the context of the clustermesh-apiserver. It could make sense to have an adapter here so that true can be always passed in that case.

joamaki replied:
Ah yes. I added a simple config struct as dependency to the cell so it's easy to specialize it.

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+			// Ignore kubernetes endpoints events
+			if ev.Key.Name == "kube-scheduler" || ev.Key.Name == "kube-controller-manager" {
+				continue
+			}
```
Nit. I'd drop this special logic, as services are only synchronized if annotated appropriately. Same below.

joamaki replied:
Dropped.

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+
+// Convert implements ClusterServiceConverter.
+func (d DefaultClusterServiceConverter) Convert(k8sService *slim_corev1.Service, eps []*k8s.Endpoints) (out *serviceStore.ClusterService, toUpsert bool) {
+	svc := serviceStore.NewClusterService(k8sService.Name, k8sService.Namespace)
```
One bit that is missing is the check whether the service is shared (i.e., `annotation.GetAnnotationShared(k8sService)`). 

If not, this function should return a `ForDeletion` service and false. Ideally, that should be performed as first operation, as it is extremely likely that the vast majority of services are not shared.

joamaki replied:
Fixed and adapted the tests

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+type DefaultClusterServiceConverter struct{}
+
+// Convert implements ClusterServiceConverter.
+func (d DefaultClusterServiceConverter) Convert(k8sService *slim_corev1.Service, eps []*k8s.Endpoints) (out *serviceStore.ClusterService, toUpsert bool) {
```
Ideally, the endpoints should be lazily retrieved here (e.g., by passing the getter function, or via an iterator), as in the vast majority of cases (i.e., if the service is not shared) they are not actually used.

joamaki replied:
Fixed!

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,264 @@
+		}
+	}
+
+	svc.Shared = annotation.GetAnnotationShared(k8sService)
```
The `Cluster` and `ClusterID` fields should be set as well.

joamaki replied:
Fixed.

giorio94 commented on operator/watchers/testdata/servicesync.txtar:
```diff
@@ -0,0 +1,151 @@
+hive/start
```
I'd suggest adding a separate test to cover changing a service from not-shared to shared, and vice versa, to ensure that it gets correctly upserted/removed (i can prepare it if you prefer).

joamaki replied:
Added these test steps. WDYT is the test file too big now and should be split?

giorio94 replied:
As you prefer honestly, it still feels readable from my point of view.

giorio94 reviewed (commented):
Thanks! This new implementation is significantly simpler and easier to follow compared to the current one. I've left a couple of initial comments inline.

joamaki committed:
k8s: Add EndpointsResourceWithIndexers

Add a constructor for endpoints resource with the option of setting
indexers. This will be used by the follow-up commit to implement
the k8s to kvstore service synchronization in the operator.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on operator/watchers/script_test.go:
```diff
@@ -0,0 +1,160 @@
+	// Due to the kvstore global variables we cannot run these tests in parallel
+	// (scripttest calls t.Parallel()). Use a mutex to serialize the test execution.
+	// Remove this once kvstore globals are removed.
+	var serializeMu lock.Mutex
```
Remove once https://github.com/cilium/hive/pull/56 is merged and bumped.

joamaki replied:
I'll do this later once it's bumped and just do a pass to remove these.

giorio94 commented on operator/watchers/script_test.go:
```diff
@@ -0,0 +1,165 @@
+			client.FakeClientCell,
+			operatorK8s.ResourcesCell,
+			cell.Config(cmtypes.DefaultClusterInfo),
+			cell.Invoke(cmtypes.ClusterInfo.InitClusterIDMax, cmtypes.ClusterInfo.Validate),
```
Nit. `cmtypes.ClusterInfo.InitClusterIDMax` touches a global variable, and it is not strictly needed here, so it could be just dropped.

giorio94 commented on operator/watchers/testdata/servicesync.txtar:
```diff
@@ -0,0 +1,183 @@
+  nodeName: nodeport-worker
+ports:
+- name: http
+  port: 80
```
Nit. Could you please change this to 8080? Just to be on the safe side and test everything.

joamaki replied:
Changed the initial port to 8080 and added a test step to change it to 9090 and check it propagates.

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,279 @@
+	svc.Cluster = d.cinfo.Name
+	svc.ClusterID = d.cinfo.ID
+	svc.Shared = true
+	svc.IncludeExternal = annotation.GetAnnotationIncludeExternal(k8sService)
```
Nit. Can be always set to `true`, as it is a per-requisite for the service to be shared.

giorio94 reviewed (approved):
Thanks! Just a couple of minor nits inline.

giorio94 reviewed (approved):
Approving again, so that it counts.

giorio94 commented on operator/watchers/service_sync.go:
```diff
@@ -0,0 +1,294 @@
+			// Skip headless services.
+			if ev.Kind != resource.Sync && isHeadless(ev.Object) {
+				continue
+			}
```
I don't think that headless services should be skipped. It is just that they should have no frontend, rather than one with the `None` address (totally missed this before).

giorio94 replied:
They were most likely dropped here in the previous service cache implementation:

https://github.com/cilium/cilium/blob/2d5639f5d8e429ebe8b6e917356c507036ec5da1/pkg/k8s/service.go#L199-L214

joamaki replied:
Ah you're right. `watchers/service.go` skips them but shouldn't do that here.
